javascript/ql/src/semmle/javascript/security/dataflow/ExceptionXss.qll

```diff
@@ -0,0 +1,86 @@
+   */
+  predicate canDefinitelyNotThrowSensitiveInformation(DataFlow::Node node) {
+    node = any(DataFlow::CallNode call | call.getCalleeName() = "getElementById").getAnArgument()
+    or
+    node = any(DataFlow::CallNode call | call.getCalleeName() = "indexOf").getAnArgument()
+    or
+    node = any(DataFlow::CallNode call | call.getCalleeName() = "stringify").getAnArgument()
+    or
+    node = DataFlow::globalVarRef("console").getAMemberCall(_).getAnArgument()
+  }
+
+  /**
+   * Holds if `node` can possibly cause an exception containing sensitive information to be thrown.
+   */
+  predicate canThrowSensitiveInformation(DataFlow::Node node) {
+    not canDefinitelyNotThrowSensitiveInformation(node) and 
+    (
+      // in the case of reflective calls the below ensures that both InvokeNodes have no known callee.
+      forex(DataFlow::InvokeNode call | node = call.getAnArgument() | not exists(call.getACallee()))
+      or
+      node.asExpr().getEnclosingStmt() instanceof ThrowStmt
+    )
+  }
+
+  /**
+   * A FlowLabel representing tainted data that has not been thrown in an exception.
+   * In the js/xss-through-exception query data-flow can only reach a sink after
+   * the data has been thrown as an exception, and data that has not been thrown
+   * as an exception therefore has this flow label, and only this flow label, associated with it.
+   */
+  class NotYetThrown extends DataFlow::FlowLabel {
+    NotYetThrown() { this = "NotYetThrown" }
+  }
+
+  /**
+   * A taint-tracking configuration for reasoning about XSS with possible exceptional flow.
+   * Flow labels are used to ensure that we only report taint-flow that has been thrown in
+   * an exception.
+   */
+  class Configuration extends TaintTracking::Configuration {
+    Configuration() { this = "ExceptionXss" }
+
+    override predicate isSource(DataFlow::Node source, DataFlow::FlowLabel label) {
+      source instanceof XSS::Shared::Source and label instanceof NotYetThrown
+    }
+
+    override predicate isSink(DataFlow::Node sink, DataFlow::FlowLabel label) {
+      sink instanceof XSS::Shared::Sink and not label instanceof NotYetThrown
+    }
+
+    override predicate isSanitizer(DataFlow::Node node) { node instanceof XSS::Shared::Sanitizer }
+
+    override predicate isAdditionalFlowStep(
+      DataFlow::Node pred, DataFlow::Node succ, DataFlow::FlowLabel inlbl,
+      DataFlow::FlowLabel outlbl
+    ) {
+      inlbl instanceof NotYetThrown and (outlbl.isTaint() or outlbl instanceof NotYetThrown) and
+      succ = pred.asExpr().getThrowsToNode() and
+      canThrowSensitiveInformation(pred)
+      or
+      // All the usual taint-flow steps apply on data-flow before it has been thrown in an exception.
+      this.isAdditionalFlowStep(pred, succ) and inlbl instanceof NotYetThrown and outlbl instanceof NotYetThrown
+      or
+      // We taint an object deep if it happens before an exception has been thrown.
```

Comment:
As usual, this seems dangerous in combination with our usual rule that taint flows from `x` to `x.f`. At the moment this won't happen, I believe, due to the previous disjunct being broken, but once you actually make `NotYetThrown`-labelled data flow from `x` to `x.f` for any `f` (as the comment on line 80 suggests) this might become problematic.

Reply:
You're right, that does look dangerous. 
It originates from me wanting to support the following pattern: 

```JavaScript
try {
  unknown({foo: taintedValue});
} catch(e) {
  sink(e);
}
```

We could drop support for that pattern and just remove that line?

Reply:
If it isn't important for flagging the CVE then that seems like the right thing to do. The usual way to solve this problem is with another flow label, but we can burn that bridge when we get to it.